### PR TITLE
Python Sol for Hamiltonian Flights in DP Bitmasks Module

### PR DESCRIPTION
**Pull request by CryoJS**
### Description

Added a Python solution to Hamiltonian Flights, content/4_Gold/DP_Bitmasks.mdx.

Let me know if there are any optimizations to be made because I can't find any. This seems to be the fastest solution for this time complexity, and to get a faster solution requires a better time complexity (not completely sure if that exists / still uses Bitmask DP).

### Submissions

Python3: https://cses.fi/problemset/result/11270874/ (TLE, optimized the same as C++ code)
PyPy3: https://cses.fi/problemset/result/11270928/ (TLE, slower: worse case not including TLE is 0.06s)
My Python3: https://cses.fi/problemset/result/11270851/ (my own solution, somewhat different than C++ solution)

My original unoptimized _strangely done_ Python solution takes 0.04 worst case (not including TLE) which is slightly slower than the optimized Python solution with a worse case not including TLE of 0.02s. The TLE is presumably due to Python's constant factor.

---

_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [X] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/4_Gold/DP_Bitmasks.mdx [hunk 1]
```diff
@@ -124,6 +124,52 @@ int main() {
 ```
 
 </CPPSection>
+<PySection>
+
+<Warning>
+
+Due to Python's large constant factor, the same C++ solution that is accepted TLEs in Python.
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Bitmasks.mdx:131]:
```suggestion
Due to Python's large constant factor, the below solution TLEs on a couple test cases.
```

File: content/4_Gold/DP_Bitmasks.mdx [hunk 2]
```diff
@@ -124,6 +124,52 @@ int main() {
...
+
+```py
+import sys
+
+input = sys.stdin.readline
+
+MOD = 10**9 + 7
+
+n, m = map(int, input().strip().split())
+dp = [[0 for _ in range(n)] for _ in range(1 << n)]  # Cities indexed from 0 to n-1
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Bitmasks.mdx:143]:
```suggestion
dp = [[0 for _ in range(n)] for _ in range(1 << n)]
```

File: content/4_Gold/DP_Bitmasks.mdx [hunk 3]
```diff
@@ -124,6 +124,52 @@ int main() {
...
+
+adj = [[] for _ in range(n)]
+
+for i in range(m):
+	a, b = map(int, input().strip().split())
+	adj[b - 1].append(a - 1)
+
+for cities in range(1, 1 << n):
+	# only consider cities that include the first city
+	if cities & 1:
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Bitmasks.mdx:154]: uh to reduce indentation you can invert the condition & use a continue

File: content/4_Gold/DP_Bitmasks.mdx [hunk 4]
```diff
@@ -124,6 +124,52 @@ int main() {
...
+		# only consider subsets with the last city if it's the full subset
+		if cities & (1 << n - 1) and cities != (1 << n) - 1:
+			continue
+
+		# loop through possible ending cities
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Bitmasks.mdx:159]: wait this is just tsp right

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
peak

File: content/4_Gold/DP_Bitmasks.mdx (hunk 1 above)
Comment (ryanchou-dev) [reply content/4_Gold/DP_Bitmasks.mdx:131]: i think the solution below doesn't tle, but is a little different than the c++ ver.

**Review by ryanchou-dev — APPROVED** [review]
after kevin's changes -- looks great!

File: content/4_Gold/DP_Bitmasks.mdx (hunk 4 above)
Comment (CryoJS) [reply content/4_Gold/DP_Bitmasks.mdx:159]: Yep

**Comment by CryoJS** [comment]
It's a small change from the original, but the most recent iteration tested on CSES:
https://cses.fi/problemset/result/11287397/

**Comment by freakin23** [comment]
Is it just me or are you ignoring my suggested changes

**Comment by SansPapyrus683** [comment]
??? my guy you haven't even suggested any

**Comment by freakin23** [comment]
> ??? my guy you haven't even suggested any (check if there's some yellow "pending" text)
Yes
Is there some approval i need for that ?

**Comment by SansPapyrus683** [comment]
no you just have to go back to the review page and finish your review

File: content/4_Gold/DP_Bitmasks.mdx [hunk 5]
```diff
@@ -124,6 +124,52 @@ int main() {
...
+
+MOD = 10**9 + 7
+
+n, m = map(int, input().strip().split())
+dp = [[0 for _ in range(n)] for _ in range(1 << n)]  # Cities indexed from 0 to n-1
+dp[1][0] = 1
+
+adj = [[] for _ in range(n)]
+
+for i in range(m):
```
Comment (freakin23) [inline content/4_Gold/DP_Bitmasks.mdx:148]:
```suggestion
for _ in range(m):
```

**Comment by freakin23** [comment]
> no you just have to go back to the review page and finish your review

Now I get it. Thanks for the help.

File: content/4_Gold/DP_Bitmasks.mdx [hunk 6]
```diff
@@ -124,6 +124,54 @@ int main() {
...
+	if not (cities & 1):
+		continue
+
+	# only consider subsets with the last city if it's the full subset
+	if cities & (1 << n - 1) and cities != (1 << n) - 1:
+		continue
+
+	# loop through possible ending cities
+	for end in range(1, n):
+		if cities & (1 << end):
```
Comment (SansPapyrus683) [inline content/4_Gold/DP_Bitmasks.mdx:163]: could you do the same condition invertion here

**Review by freakin23 — APPROVED** [review]

**Comment by freakin23** [comment]
the submission links that you have attached are not working, it is better to attach screenshots of your submission.

File: content/4_Gold/DP_Bitmasks.mdx (hunk 6 above)
Comment (CryoJS) [reply content/4_Gold/DP_Bitmasks.mdx:163]: Got it, I will also do the same on another similar line below that too

**Review by SansPapyrus683 — APPROVED** [review]

**Pull request merged**